ibis/expr/rules.py

```diff
@@ -424,11 +424,32 @@ def table(arg, *, schema=None, **kwargs):
 
 @validator
 def column_from(name, column, *, this):
-    if not isinstance(column, (str, int)):
-        raise com.IbisTypeError(
-            f"value must be an int or str, got {type(column).__name__}"
-        )
-    return getattr(this, name)[column]
+    if not hasattr(this, name):
+        raise com.IbisTypeError(f"Could not get table {name} from {this}")
+    table = getattr(this, name)
+    if isinstance(column, ir.AnyColumn):
+        if not column.has_name():
+            raise com.IbisTypeError(f"Passed column {column} has no name")
+
+        maybe_column = column.get_name()
+        try:
+            if column.equals(table[maybe_column]):
+                return column
+            else:
+                raise com.IbisTypeError(
+                    f"Passed column is not a column in {table}"
+                )
+        except com.IbisError:
+            raise com.IbisTypeError(
```

Comment:
Can we just let the `IbisError` propagate?

Reply:
This causes large test failures.

Reply:
https://github.com/ibis-project/ibis/runs/4562427238?check_suite_focus=true as an example.

Reply:
Got it.

Reply:
We could change the `suppress` call here to accept `IbisError` as well if we wanted to not let it propagate: https://github.com/ibis-project/ibis/blob/7da815cf725fa2001d1608922ca261ac2a74e788/ibis/expr/rules.py#L98. 

I think that would be cleaner, but I don't know of the original intent of only using `IbisTypeError` here.